### ISSUE TITLE
Which model to use for decoding/evaluation?

### ISSUE THREAD

**Issue by wasiahmad** [kenchan0226/keyphrase-generation-rl#6]
Once training is finished, I see lot of model files but not sure which one to use as the final one. For example, if I have the following two models, which one I should use?

```
openkp.ml.one2many.cat.copy.bi-directional.epoch=2.batch=3569.total_batch=12000.model
openkp.ml.one2many.cat.copy.bi-directional.epoch=2.batch=7569.total_batch=16000.model
```

Also, curious what the `batch` and `total_batch` values mean here? I assume we need to use the model that is resulted from the latest `epoch`. So, I use model with `epoch=2` than `epoch=1`.

**Comment by avinsit123** [comment]
Batch defines the no.of samples on that the model has been trained on in the current epoch. And total_batch indicates total no.of samples it has been trained upon including all the previous epochs.
Also once you train , every time it saves the model it showcases the test and validation loss along with where it is saved you can choose the best model from there.

**Comment by wasiahmad** [comment]
I am confused. I have a few questions. 

1. I have around 100k training examples and I have set the batch size to 16. So, what does `epoch=2.batch=3569` or `epoch=2.batch=7569` mean? 
2. Why there are two files for the same epoch? Do you perform validation more than once during 1 epoch? After how many mini-batches, you perform the validation?
3. Do you save a model even if it doesn't result in improved performance from the last checkpoint? Is this assumption wrong that "the latest checkpoint is the best model"? I have the following log.

```
10/24/2019 20:19:11 [INFO] train_ml: Saving checkpoint to model/catSeqD/openkp.ml.one2many.cat.copy.orthogonal.target_encode.bi-directional.epoch=2.batch=7569.total_batch=16000.model
INFO:root:Epoch: 2; batch idx: 7569; total batches: 16000
10/24/2019 20:19:11 [INFO] train_ml: Epoch: 2; batch idx: 7569; total batches: 16000
INFO:root:avg training ppl: 4.693; avg validation ppl: 4.131; best validation ppl: 4.131
10/24/2019 20:19:11 [INFO] train_ml: avg training ppl: 4.693; avg validation ppl: 4.131; best validation ppl: 4.131
INFO:root:avg training loss: 1.546; avg validation loss: 1.418; best validation loss: 1.418
10/24/2019 20:19:11 [INFO] train_ml: avg training loss: 1.546; avg validation loss: 1.418; best validation loss: 1.418
Epoch 3; batch: 3138; total batch: 20000
Enter check point!
Valid loss does not drop
INFO:root:Epoch: 3; batch idx: 3138; total batches: 20000
10/24/2019 20:52:23 [INFO] train_ml: Epoch: 3; batch idx: 3138; total batches: 20000
INFO:root:avg training ppl: 4.147; avg validation ppl: 4.370; best validation ppl: 4.131
10/24/2019 20:52:23 [INFO] train_ml: avg training ppl: 4.147; avg validation ppl: 4.370; best validation ppl: 4.131
INFO:root:avg training loss: 1.422; avg validation loss: 1.475; best validation loss: 1.418
10/24/2019 20:52:23 [INFO] train_ml: avg training loss: 1.422; avg validation loss: 1.475; best validation loss: 1.418
Epoch 3; batch: 7138; total batch: 24000
Enter check point!
Valid loss does not drop
INFO:root:Epoch: 3; batch idx: 7138; total batches: 24000
10/24/2019 21:25:40 [INFO] train_ml: Epoch: 3; batch idx: 7138; total batches: 24000
INFO:root:avg training ppl: 3.896; avg validation ppl: 4.172; best validation ppl: 4.131
10/24/2019 21:25:40 [INFO] train_ml: avg training ppl: 3.896; avg validation ppl: 4.172; best validation ppl: 4.131
INFO:root:avg training loss: 1.360; avg validation loss: 1.428; best validation loss: 1.418
10/24/2019 21:25:40 [INFO] train_ml: avg training loss: 1.360; avg validation loss: 1.428; best validation loss: 1.418
Epoch 4; batch: 2707; total batch: 28000
Enter check point!
Valid loss does not drop
```

Can you tell me what does the following mean?

```
Enter check point!
Valid loss does not drop
```

**Comment by avinsit123** [comment]
Sure Let me answer all of them one by one .

1.  Consider this log which will give you a good idea of "batch" and "total_batch" 

> INFO:root:======================  Start Training  =========================
> 06/28/2019 11:47:48 [INFO] train_ml: ======================  Start Training  =========================
> Epoch 1; batch: 0; total batch: 0  
> Epoch 1; batch: 4000; total batch: 4000 
> Epoch 1; batch: 8000; total batch: 8000
> Epoch 1; batch: 12000; total batch: 12000
> Epoch 1; batch: 16000; total batch: 16000
> Epoch 1; batch: 20000; total batch: 20000
> Epoch 1; batch: 24000; total batch: 24000
> Epoch 1; batch: 28000; total batch: 28000
> Epoch 1; batch: 32000; total batch: 32000
> Epoch 1; batch: 36000; total batch: 36000
> Epoch 1; batch: 40000; total batch: 40000
> Epoch 1; batch: 44000; total batch: 44000
> Epoch 1; batch: 48000; total batch: 48000
> Epoch 1; batch: 52000; total batch: 52000
> Epoch 1; batch: 56000; total batch: 56000
> Epoch 1; batch: 60000; total batch: 60000
> Epoch 2; batch: 2573; total batch: 64000
> Enter check point!
> Valid loss drops
> INFO:root:Saving checkpoint to model/kp20k.ml.one2many.cat.copy.bi-directional.20190628-114655/kp20k.ml.one2many.cat.copy.bi-directional.epoch=2.batch=2573.total_batch=64000.model

**The total no.of batches for my dataset was 61427.**
As you see in the first epoch, no.of "batch" is equal to the "total_batch". In the 2nd batch it decreases to 2573. This because the total no.of batches run till now is 64000, but in the 2nd epoch only 2573 batches ( 64000 % 61427) have been run so batch indicates the total no.of batches run since start of the epoch.

2. Yes Validation is done more than once for each epoch. It is done after every 4000 batches.
3. No, the model is saved only if the valid loss drops. Thus your assumption that the latest model is the best model is the right one.

**Comment by kenchan0226** [comment]
Thanks Avinash! You accurately answered those questions.
